Use visual studio to setup the local `resources` tree

kcgen opened the pull request:
This makes the visual studio solution comparable to the meson when it comes to seting the `resources` tree in the build area.  

Thanks to @NicknineTheEagle for reporting this.

shermp commented:
Just tested this on my local VS2019, it works, but I have some concerns.

First, I can't seem to find the changes reflected it the VS solution explorer or project config anywhere. Not sure if this will be an issue or not.

Second, this is a pretty indiscriminate copy, and includes unwanted files such as shell scripts, and the `meson.build` files.

kcgen commented:
> First, I can't seem to find the changes reflected it the VS solution explorer or project config anywhere. Not sure if this will be an issue or not.

Just figuring this out took many different attempts at variations.. and I had actually given up and added a copy step inside the YAML file instead, which does the job just the same, but the downside is that users compiling at home wouldn't get the automatic copy of the `resources/` into their build target area.

So; even though this is true - ie: VS GUI doesn't have widgets or text boxes showing this step, the question is, what can we do about it? (The Microsoft notes actually instruct to "edit the XML file", and not use the UI if you want to work with the sytax I added).


> Second, this is a pretty indiscriminate copy, and includes unwanted files such as shell scripts, and the meson.build files.

Yes.  I also don't have a perfect solution for that either.

There are some options:

1. Add per-file item and copy XML statements inside the primary `.vcproj` and the unit tests `.vcproj` files.  By being explicit, we could omit the extra files present that get bulk-copied.

2. Option 1 is brutal to me, because now we've got the names of these files hidden away inside two massive project-wide XML file (with duplicate information) that we have to remember to modify if we want to adjust the resource files. So a second option would be figuring out how to read a list of just the deliverable files sitting inside each directory, exactly like we've got for Meson.   But unfortunately I haven't come across how to read files dynamically into a vcproj.

3. There are provisions in the vcproj to calls an OS command after the build is complete, so we could use that to call a Python script that reads the deliverable file list variable `resources = ["a", "b", ... ]` and perform the per-file copy. But for this, we would need Python and figure out how to call that (or would need to parse that file list content).

4. We could helicopter up and eliminate the problem such that the bulk-copy _cannot_ include extraneous files (well, perhaps it would still contain the meson.build files, beacuse Meson is declarative and we can't do things like copy *.* in it).  But the idea is that we move the DOCS/ and LICENSE/ into a separate metadata directory that describing 3rd party content (a bit like a compliance bucket).

Maybe there are other options?

In the end, the bulk-copy approach was a one-time addition (that won't require further maintenance), in a secondary build system (ie: the only build system we deliver from is Meson and MSYS2); because Visual Studio can't even build a full-featured Staging binary.  So in the end, the couple extra kilobytes and could tiny DOC and text files is close to a no-op compared to the alternatives, which are far more onerous.

shermp commented:
Hmm, is this going to be an issue? 

https://docs.microsoft.com/en-us/cpp/build/reference/vcxproj-files-and-wildcards?view=msvc-170

kcgen commented:
> Hmm, is this going to be an issue?
> 
> https://docs.microsoft.com/en-us/cpp/build/reference/vcxproj-files-and-wildcards?view=msvc-170

The syntax used to bulk-copy `contrib/resources` (and retain just the `resources/` directory in the build-root (and drop `contrib` doesn't use name & extension wildcards:

```xml
        <Link>..\contrib\resources\%(RecursiveDir)\%(Filename)%(Extension)</Link>
        <TargetPath>resources\%(RecursiveDir)\%(Filename)%(Extension)</TargetPath>
        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
```

shermp commented:
It's the `Include` attribute I'm worried about. That looks like a wildcard to me!

`Include="..\contrib\resources\**"`

kcgen commented:
It's from their 2022 docs, and specifically the `**` maps to the `RecursiveDir`, `Filename` and `Extension` built-ins.

https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2022

kcgen commented:
Regardless though, this is actually working in Visual Studio 2022.  We could preemptively throw it out (making our lives more difficult, I guess?).

Heck.. even the old syntax from VS2005 was still working (and that's using `*.*` file-pattern wildcards). .. but that approach didn't retain the `resources/` target subdirectory.

shermp commented:
> It's from their 2022 docs, and specifically the `**` maps to the `RecursiveDir`, `Filename` and `Extension` built-ins.
> 
> https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2022

Yeah, MSBuild certainly supports it, I'm not sure the IDE supports it for C++ projects though. Note, the link I provided was specifically for C++ projects. MSBuild is also used for dotnet etc as well, which may not have the same limitations.

kcgen commented:
> I'm not sure the IDE supports it for C++ projects 

Ahh; that explains why the help docs described having to open the XML directly, I guess.

shermp commented:
This is the main warning I'm worried about:

> If a .vcxproj project file that includes these constructs gets loaded in the IDE, the project may seem to work at first. However, issues are likely as soon as the project is modified by Visual Studio and then saved on disk. You may experience random crashes and undefined behavior.

kcgen commented:
And I guess that's even against their current 2022 VS release? (I like how they put a showstoping bug like that back on the user's plate instead of saying "we have a severe bug that can't handle wildcards inside XML; we will disable saving project files if this is detected"). 

The saving grace for us is that the repos project file is only modified on PR rwviyand merge..  it I suppose users could still modify their repo's copy and introduce the problem locally. 

Maybe we should do number 4, plus switch to an xcopy call (in the project file) that just grabs `contrib/resources`, and never uses `*`?  There are probably built-ins to get the output dir.

shermp commented:
> And I guess that's even against their current 2022 VS release?

Looks like it.

You're supposed to add files to the project via the IDE. Of course, VS can't add an entire directory tree. That would be far too convenient...

![image](https://user-images.githubusercontent.com/14854761/161008356-ded55b9f-ba72-4d52-a4be-7df5085b73e6.png)

shermp commented:
Thinking about it, calling a post build event is probably the easiest solution, and call xcopy. Looks like variables are supported, eg from the docs:
```xml
<PostBuildEvent>
    <Command>copy $(ProjectDir)$(Configuration)\$(TargetFileName) $(ProjectDir)$(Configuration)\copyOfMyproject.exe</Command>
    <Message>Making a copy of myproject.exe</Message>
  </PostBuildEvent>
```

kcgen commented:
I've only ever modified the XML directly, and I wasn't planning on adding any of the `contrib/` tree content to be shown in the editor (these are mostly binaries anyway.. I'm trying to keep them away from editors, search patterns, spell checkers, and so on. )

But if users of visual studio normally expect binaries by listed in the vxproj, is this something you can add?

I still think we need a one-shot xcopy call if we want to avoid use of wildcards.

kcgen commented:
Perfect @shermp ; thanks for that snippet. 👍 

Will add that tomorrow.

shermp commented:
You should be able to add the following as the last child of each `<ItemDefinitionGroup>`.

```xml
<PostBuildEvent>
    <Command>xcopy ..\contrib\resources $(OutDir)resources /s /i /y</Command>
</PostBuildEvent>
<PostBuildEvent>
    <Message>Copy resources to output directory</Message>
</PostBuildEvent>
```

(Adjust the xcopy command to suit, although the above seems to work ok)

NicknineTheEagle commented:
> Regardless though, this is actually working in Visual Studio 2022. We could preemptively throw it out (making our lives more difficult, I guess?).
> 
> Heck.. even the old syntax from VS2005 was still working (and that's using `*.*` file-pattern wildcards). .. but that approach didn't retain the `resources/` target subdirectory.

I should note that I'm still on VS 2019.

kcgen commented:
Thanks @shermp and for the heads up @NicknineTheEagle.
Update inbound.

kcgen commented:
Getting failures in the release build. On try `#4` to work around it;
 - https://stackoverflow.com/questions/22730003/msb3073-command-exited-with-code-9009
 - https://stackoverflow.com/questions/19914339/error-msb3073-how-do-i-fix-this

![2022-03-31_11-08](https://user-images.githubusercontent.com/1557255/161121675-7a10d55e-bfa3-42d9-a22d-e92bd23924ae.png)

shermp commented:
The actual error is further up:

```
PostBuildEvent:
         Copy resources to output directory
         xcopy ..\contrib\resources Win32\Release\resources /s /i /y
         :VCEnd
         'xcopy' is not recognized as an internal or external command,
         operable program or batch file.
```
WTF GitHub?

kcgen commented:
Thanks @shermp . 

Indeed, these work for the debug builds:
```xml
        <Command>xcopy ..\contrib\resources "$(OutDir)\resources" /s /i /y</Command>
```

Let's try the full path for the release variants:
```xml
        <Command>%systemroot%\System32\xcopy ..\contrib\resources "$(OutDir)\resources" /s /i /y</Command>
```

kcgen commented:
I think we're on the home stretch, after 13 attempts (and even more stack-oveflow and Microsoft dev forums later).

kcgen commented:
- It passes tests (thanks @NicknineTheEagle )
 - It avoids using wildcards (thanks @shermp)
 - It can be managed inside the VisualStudio config UI (thanks @shermp)
 - It gets the job done for those compiling using Visual Studio and running straight from the build area.
 - It puts the resources in the build directory so the `create-package.sh` bundles the resources. 

I think it's good enough as we can get; so let's merge and carry on.

If there are remaining issues, we can follow up from there.
